oml/datasets/audios.py

```diff
@@ -0,0 +1,353 @@
+import random
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Union
+
+import numpy as np
+import pandas as pd
+import torch
+import torchaudio
+from numpy.typing import NDArray
+from torch import BoolTensor, FloatTensor, LongTensor
+from torchaudio.transforms import MelSpectrogram, Resample
+
+from oml.const import (
+    AUDIO_EXTENSIONS,
+    BLACK,
+    CATEGORIES_COLUMN,
+    DEFAULT_DURATION,
+    DEFAULT_MELSPEC_PARAMS,
+    DEFAULT_NUM_CHANNELS,
+    DEFAULT_SAMPLE_RATE,
+    INDEX_KEY,
+    INPUT_FRAME_OFFSET_KEY,
+    INPUT_TENSORS_KEY,
+    IS_GALLERY_COLUMN,
+    IS_QUERY_COLUMN,
+    LABELS_COLUMN,
+    LABELS_KEY,
+    PATHS_COLUMN,
+    SEQUENCE_COLUMN,
+    TColor,
+)
+from oml.interfaces.datasets import (
+    IBaseDataset,
+    ILabeledDataset,
+    IQueryGalleryDataset,
+    IVisualizableDataset,
+)
+from oml.utils.audios import visualize_audio, visualize_audio_html
+
+
+class AudioBaseDataset(IBaseDataset, IVisualizableDataset):
+    """
+    The base class that handles audio specific logic.
+
+    """
+
+    def __init__(
+        self,
+        paths: List[Union[str, Path]],
+        dataset_root: Optional[Union[str, Path]] = None,
+        extra_data: Optional[Dict[str, Any]] = None,
+        sr: int = DEFAULT_SAMPLE_RATE,
+        num_seconds: float = DEFAULT_DURATION,
+        num_channels: int = DEFAULT_NUM_CHANNELS,
+        input_tensors_key: str = INPUT_TENSORS_KEY,
+        input_frame_offset_key: str = INPUT_FRAME_OFFSET_KEY,
```

Comment:
I'm not sure we need this key and suggest to remove the key and the constant `INPUT_FRAME_OFFSET_KEY`. Modality related information should not appear outside of the dataset.

The purpose of `extra_data` is to store some external information provided from outside (usually by user in runtime or when we have extra columns in the dataframe). To be honest, I did not understand from where we take this offsets, because there are no offsets in the dataframe.

By the way? Do you need to be used later, for example, in the model?

Reply:
i thought a lot on this question, so:
1. I changed naming `frame_offset -> start_time`, so it would represent not the number of frames, but the number of seconds.
2. Instead of adding explicitly `start_time`(prev. `frame_offset`) column as dataset argument, I added bool arg `use_random_start`, which works as following:
  - if True, there will be random starts and no `start_time` will be applied
  - If False, it will search for `start_time` in `extra_data`, and if it fails, there will be Exception.

Reply:
Hm, what is the general purpose of random start? Is that type of training augmentation?
Do I understand correctly, that we never use such option in the validation, because we want to have reproducible results?

Reply:
yeah, it is basically some kind of augmentation and yeah, we'll never use it in the validation. So there can be several situations:
- user defined start times in csv/extra data, so these values will be used both on training and validation
- user did not defined offsets in csv/extra data, so on the training section audio may have random start or start on beginning (0) and on validation only on beginning too (I'll add this logic).

Essentially, we need these start_times/random_start because audios are mostly longer than it is required for the model, so we need to decide how to trim it. A lot of datasets come with the `start_time` of the training/validation segment, so you just cut it from the particular point, but sometimes they dont, so we need to trim them mostly randomly.

Reply:
Got it.

If there is a common practice when people make `random_start` the part of training transforms (augmentations)? There is an image related analogue: cropping images during the training (CropResize). If so, could we also move this to the augmentations?

As for the validation, it's look fine to me.

I'm also worried about potential conflict between `start_times` and `use_random_start_time`. Let's raise an error if 
`use_random_start_time=True`, but we also had `START_TIME_COLUMN` in our dataframe.

Reply:
Another thing: we should add and example with `START_TIME_COLUMN` to the mock dataset, as we did it for images: `df_with_bboxes.csv`. First, it will serve as example, second, we can use it in tests.

Reply:
done